ledger/src/blockstore/blockstore_purge.rs

```diff
@@ -138,96 +139,192 @@ impl Blockstore {
         self.run_purge_with_stats(from_slot, to_slot, purge_type, &mut PurgeStats::default())
     }
 
+    /// Purges all columns relating to `slot`.
+    ///
+    /// Additionally we cleanup the parent of `slot`, by clearing `slot` from
+    /// the parent's `next_slots`. We reinsert an orphaned `slot_meta` for `slot`
+    /// that preserves `slot`'s `next_slots`. This ensures that `slot`'s fork is
+    /// replayable upon repair of `slot`.
+    pub(crate) fn purge_slot_cleanup_chaining(&self, slot: Slot) -> Result<bool> {
+        let Some(mut slot_meta) = self.meta(slot)? else {
+            return Err(BlockstoreError::SlotUnavailable);
+        };
+        let mut write_batch = self.db.batch()?;
+
+        let columns_purged = self.purge_range(&mut write_batch, slot, slot, PurgeType::Exact)?;
+
+        if let Some(parent_slot) = slot_meta.parent_slot {
+            let parent_slot_meta = self.meta(parent_slot)?;
+            if let Some(mut parent_slot_meta) = parent_slot_meta {
+                // .retain() is a linear scan; however, next_slots should
+                // only contain several elements so this isn't so bad
+                parent_slot_meta
+                    .next_slots
+                    .retain(|&next_slot| next_slot != slot);
+                write_batch.put::<cf::SlotMeta>(parent_slot, &parent_slot_meta)?;
+            } else {
+                error!(
+                    "Parent slot meta {} for child {} is missing  or cleaned up.
+                       Falling back to orphan repair to remedy the situation",
+                    parent_slot, slot
+                );
+            }
```

Comment:
This is a pretty large BP to consider for v1.17 & v1.18; most of the change was a refactor to clean things up. I think we could get away with BP'ing much less, namely, just this bit to print an error instead of doing the following expect that is currently in the v1.17/18 branches:
```rust
.expect("Unconfirmed slot should have had parent slot set");
```

Reply:
agreed, i think we can just error instead of expect. the atomic write can wait for master

Reply:
Right. And thinking out loud, suppose that `C` is the child of `P`. If we purge `C` but can't find `P`'s meta, we'll repopulate a meta for `P` when `C` is repaired / reinserted

Reply:
yep that's correct